operator/controllers/keptnappversion/controller.go

```diff
@@ -0,0 +1,207 @@
+	Scheme   *runtime.Scheme
+	Log      logr.Logger
+	Recorder record.EventRecorder
+	Tracer   trace.Tracer
+	Meters   common.KeptnMeters
+}
+
+//+kubebuilder:rbac:groups=lifecycle.keptn.sh,resources=keptnappversions,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=lifecycle.keptn.sh,resources=keptnappversions/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=lifecycle.keptn.sh,resources=keptnappversions/finalizers,verbs=update
+//+kubebuilder:rbac:groups=lifecycle.keptn.sh,resources=keptnworkloadinstances/status,verbs=get;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the KeptnAppVersion object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
+func (r *KeptnAppVersionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.Log = log.FromContext(ctx)
+	r.Log.Info("Searching for Keptn App Version")
+
+	appVersion := &klcv1alpha1.KeptnAppVersion{}
+	err := r.Get(ctx, req.NamespacedName, appVersion)
+	if errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+
+	if err != nil {
+		r.Log.Error(err, "App Version not found")
+		return reconcile.Result{}, fmt.Errorf("could not fetch KeptnappVersion: %+v", err)
+	}
+
+	if !appVersion.IsStartTimeSet() {
+		// metrics: increment active app counter
+		r.Meters.AppActive.Add(ctx, 1, appVersion.GetActiveMetricsAttributes()...)
+		appVersion.SetStartTime()
+	}
+
+	traceContextCarrier := propagation.MapCarrier(appVersion.Annotations)
+	ctx = otel.GetTextMapPropagator().Extract(ctx, traceContextCarrier)
+
+	ctx, span := r.Tracer.Start(ctx, "reconcile_app_version", trace.WithSpanKind(trace.SpanKindConsumer))
+	defer span.End()
+
+	semconv.AddAttributeFromAppVersion(span, *appVersion)
+
+	phase := common.PhaseAppPreDeployment
+	if !appVersion.IsPreDeploymentSucceeded() {
+		r.Log.Info(fmt.Sprintf("%s Tasks not finished", phase.LongName))
+		if appVersion.IsPreDeploymentFailed() {
+			r.recordEvent(phase, "Warning", appVersion, "Failed", "has failed")
+			return ctrl.Result{Requeue: true, RequeueAfter: 60 * time.Second}, nil
+		}
+		state, err := r.reconcilePrePostDeployment(ctx, appVersion, common.PreDeploymentCheckType)
+		if err != nil {
+			span.SetStatus(codes.Error, err.Error())
+			r.recordEvent(phase, "Warning", appVersion, "ReconcileErrored", "could not get reconciled")
+			return ctrl.Result{Requeue: true}, err
+		}
+		if state.IsSucceeded() {
+			r.recordEvent(phase, "Normal", appVersion, "Succeeded", "has succeeded")
+		} else {
+			r.recordEvent(phase, "Warning", appVersion, "NotFinished", "has not finished")
```

Comment:
does this mean we keep sending warning events while the pre deployment tasks are running?

Reply:
I think these were introduced for debugging reasons

Reply:
Should we get rid of them @thschue ?

Reply:
No, but we should make them normal ...